README.md
# pcm转化工具

## 使用

```
/**
 *
 * @param samples: [Array: ArrayBuffet] pcm数据
 * @param sampleRateTmp [number] 采样率 8000/16000
 * @param sampleBits [number] 位数 8/16/32
 * @param channelCount [Number] 声道,单声道传1,多声道传2
 * return {
 *  buffer, // wave的ArrayBuffer
    waveArray: // wave的普通数组
    blobUrl: // wave的blob url用户音频播放
 * }
 */
 const {blobUrl} = pcmTransfer(samples, sampleRateTmp, sampleBits, channelCount);
```


changelog.md
### 1.0.2 2019-12-4

+ 修复代码格式
+ 添加example

### 1.0.0 2019-12-4

+ 初步完成功能


lib/pcm-transfer.js
!function(e,t){"object"==typeof exports&&"object"==typeof module?module.exports=t():"function"==typeof define&&define.amd?define("pcmTransfer",[],t):"object"==typeof exports?exports.pcmTransfer=t():e.pcmTransfer=t()}(window,(function(){return function(e){var t={};function r(n){if(t[n])return t[n].exports;var o=t[n]={i:n,l:!1,exports:{}};return e[n].call(o.exports,o,o.exports,r),o.l=!0,o.exports}return r.m=e,r.c=t,r.d=function(e,t,n){r.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:n})},r.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},r.t=function(e,t){if(1&t&&(e=r(e)),8&t)return e;if(4&t&&"object"==typeof e&&e&&e.__esModule)return e;var n=Object.create(null);if(r.r(n),Object.defineProperty(n,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var o in e)r.d(n,o,function(t){return e[t]}.bind(null,o));return n},r.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return r.d(t,"a",t),t},r.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},r.p="",r(r.s=0)}([function(e,t,r){"use strict";r.r(t),t.default=function(e,t,r,n){var o=(e=function(e,t){return e.byteLength?e:8===t?new Int8Array(e).buffer:16===t?new Int16Array(e).buffer:new Int32Array(e).buffer}(e,r)).byteLength,f=new ArrayBuffer(44+o),u=new DataView(f);function a(e,t,r){for(var n=0;n<r.length;n++)e.setUint8(t+n,r.charCodeAt(n))}var i=0;a(u,i,"RIFF"),i+=4,u.setUint32(i,36+o,!0),a(u,i+=4,"WAVE"),a(u,i+=4,"fmt "),i+=4,u.setUint32(i,16,!0),i+=4,u.setUint16(i,1,!0),i+=2,u.setUint16(i,n,!0),i+=2,u.setUint32(i,t,!0),i+=4,u.setUint32(i,t*n*(r/8),!0),i+=4,u.setUint16(i,n*(r/8),!0),i+=2,u.setUint16(i,r,!0),a(u,i+=2,"data"),i+=4,u.setUint32(i,o,!0),i+=4;var c=[];return c=16===r?function(e,t,r){r=new Int16Array(r);for(var n=0;n<r.length;n+=1,t+=2)e.setInt16(t,r[n],!0);return new Int16Array(e.buffer)}(u,44,e):8===r?function(e,t,r){r=new Int8Array(r);for(var n=0;n<r.length;n+=1,t+=1)e.setInt8(t,r[n]);return new Int8Array(e.buffer)}(u,44,e):function(e,t,r){r=new Int32Array(r);for(var n=0;n<r.length;n+=1,t+=4)e.setInt32(t,r[n],!0);return new Int32Array(e.buffer)}(u,44,e),{buffer:u.buffer,waveArray:Array.prototype.slice.call(c),blobUrl:window.URL.createObjectURL(new Blob([u.buffer],{type:"audio/wav"}))}}}]).default}));

src/index.ts

function arrayToBuffer(array: any, sampleBits: number) {
  const isBuffer = array.byteLength;
  if (isBuffer) return array;
  if (sampleBits === 8) {
    return new Int8Array(array).buffer;
  } if (sampleBits === 16) {
    return new Int16Array(array).buffer;
  }
  return new Int32Array(array).buffer;
}

function floatTo32BitPCM(output: DataView, offset: number, input: any) {
  input = new Int32Array(input);
  for (let i = 0; i < input.length; i += 1, offset += 4) {
    output.setInt32(offset, input[i], true);
  }
  return new Int32Array(output.buffer);
}

function floatTo16BitPCM(output: DataView, offset: number, input: any) {
  input = new Int16Array(input);
  for (let i = 0; i < input.length; i += 1, offset += 2) {
    output.setInt16(offset, input[i], true);
  }
  return new Int16Array(output.buffer);
}

function floatTo8BitPCM(output: DataView, offset: number, input: any) {
  input = new Int8Array(input);
  for (let i = 0; i < input.length; i += 1, offset += 1) {
    output.setInt8(offset, input[i]);
  }
  return new Int8Array(output.buffer);
}

/**
 *
 * @param samples: [Array: ArrayBuffet] pcm数据
 * @param sampleRateTmp [number] 采样率 8000/16000
 * @param sampleBits [number] 位数 8/16/32
 * @param channelCount [Number] 声道,单声道传1,多声道传2
 * return {
 *  buffer, // wave的ArrayBuffer
    waveArray: // wave的普通数组
    blobUrl: // wave的blob url用户音频播放
 * }
 */

function pcmTransfer(samples: any,
  sampleRateTmp: number,
  sampleBits: number,
  channelCount: number) {
  samples = arrayToBuffer(samples, sampleBits);
  const dataLength = samples.byteLength;
  const buffer = new ArrayBuffer(44 + dataLength);
  const view = new DataView(buffer);

  function writeString(dv: DataView, offset: number, string: string) {
    for (let i = 0; i < string.length; i++) {
      dv.setUint8(offset + i, string.charCodeAt(i));
    }
  }
  let offset = 0;
  /* 资源交换文件标识符 */
  writeString(view, offset, 'RIFF'); offset += 4;
  /* 下个地址开始到文件尾总字节数,即文件大小-8 */
  view.setUint32(offset, /* 32 */ 36 + dataLength, true); offset += 4;
  /* WAV文件标志 */
  writeString(view, offset, 'WAVE'); offset += 4;
  /* 波形格式标志 */
  writeString(view, offset, 'fmt '); offset += 4;
  /* 过滤字节,一般为 0x10 = 16 */
  view.setUint32(offset, 16, true); offset += 4;
  /* 格式类别 (PCM形式采样数据) */
  view.setUint16(offset, 1, true); offset += 2;
  /* 通道数 */
  view.setUint16(offset, channelCount, true); offset += 2;
  /* 采样率,每秒样本数,表示每个通道的播放速度 */
  view.setUint32(offset, sampleRateTmp, true); offset += 4;
  /* 波形数据传输率 (每秒平均字节数) 通道数×每秒数据位数×每样本数据位/8 */
  view.setUint32(offset, sampleRateTmp * channelCount * (sampleBits / 8), true); offset += 4;
  /* 快数据调整数 采样一次占用字节数 通道数×每样本的数据位数/8 */
  view.setUint16(offset, channelCount * (sampleBits / 8), true); offset += 2;
  /* 每样本数据位数 */
  view.setUint16(offset, sampleBits, true); offset += 2;
  /* 数据标识符 */
  writeString(view, offset, 'data'); offset += 4;
  /* 采样数据总数,即数据总大小-44 */
  view.setUint32(offset, dataLength, true); offset += 4;
  let waveArray: any = [];
  if (sampleBits === 16) {
    waveArray = floatTo16BitPCM(view, 44, samples);
  } else if (sampleBits === 8) {
    waveArray = floatTo8BitPCM(view, 44, samples);
  } else {
    waveArray = floatTo32BitPCM(view, 44, samples);
  }
  return {
    buffer: view.buffer,
    waveArray: Array.prototype.slice.call(waveArray),
    blobUrl: window.URL.createObjectURL(new Blob([view.buffer], { type: 'audio/wav' })),
  };
}

export default pcmTransfer;
